pkg/pillar/cmd/vcomlink/vcomlink.go

```diff
@@ -0,0 +1,233 @@
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	ctx.subGlobalConfig = subGlobalConfig
+	subGlobalConfig.Activate()
+
+	for !ctx.GCInitialized {
+		log.Functionf("waiting for GCInitialized")
+		select {
+		case change := <-subGlobalConfig.MsgChan():
+			subGlobalConfig.ProcessChange(change)
+		case <-stillRunning.C:
+		}
+		ps.StillRunning(agentName, warningTime, errorTime)
+	}
+	log.Functionf("processed GlobalConfig")
+
+	// start listening on vsock
+	go startVsockServer()
+
+	for {
+		select {
+		case change := <-subGlobalConfig.MsgChan():
+			subGlobalConfig.ProcessChange(change)
+		case <-stillRunning.C:
+		}
+		ps.StillRunning(agentName, warningTime, errorTime)
+	}
+}
+
+func handleGlobalConfigCreate(ctxArg interface{}, key string,
+	statusArg interface{}) {
+	handleGlobalConfigImpl(ctxArg, key, statusArg)
+}
+
+func handleGlobalConfigModify(ctxArg interface{}, key string,
+	statusArg interface{}, oldStatusArg interface{}) {
+	handleGlobalConfigImpl(ctxArg, key, statusArg)
+}
+
+func handleGlobalConfigImpl(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	ctx := ctxArg.(*vcomLinkContext)
+	if key != "global" {
+		log.Functionf("handleGlobalConfigImpl: ignoring %s", key)
+		return
+	}
+	log.Functionf("handleGlobalConfigImpl for %s", key)
+	gcp := agentlog.HandleGlobalConfig(log, ctx.subGlobalConfig, agentName,
+		ctx.CLIParams().DebugOverride, logger)
+	if gcp != nil {
+		ctx.GCInitialized = true
+	}
+	log.Functionf("handleGlobalConfigImpl done for %s", key)
+}
+
+func handleGlobalConfigDelete(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	ctx := ctxArg.(*vcomLinkContext)
+	if key != "global" {
+		log.Functionf("handleGlobalConfigDelete: ignoring %s", key)
+		return
+	}
+	log.Functionf("handleGlobalConfigDelete for %s", key)
+	agentlog.HandleGlobalConfig(log, ctx.subGlobalConfig, agentName,
+		ctx.CLIParams().DebugOverride, logger)
+	log.Functionf("handleGlobalConfigDelete done for %s", key)
+}
+
+func startVsockServer() {
+	// XXX : this rudimentary vsock server, it still can handle multiple VMs
+	// but if it gets too complex in the future it can be improved by
+	// assigning each vm or service a unique port.
+	sock, err := unix.Socket(unix.AF_VSOCK, unix.SOCK_STREAM, 0)
+	if err != nil {
+		log.Errorf("failed to create vsock socket: %v", err)
+		return
+	}
+	defer unix.Close(sock)
+
+	addr := &unix.SockaddrVM{
+		CID:  unix.VMADDR_CID_HOST,
+		Port: vcom.HostVPort,
+	}
+	if err := unix.Bind(sock, addr); err != nil {
+		log.Errorf("failed to bind vsock socket: %v", err)
+		return
+	}
+	if err := unix.Listen(sock, backLogSize); err != nil {
+		log.Errorf("failed to listen on vsock socket: %v", err)
+		return
+	}
+	log.Noticef("Listening on vsock CID %d, port %d", addr.CID, addr.Port)
+
+	for {
+		conn, _, err := unix.Accept(sock)
+		if err != nil {
+			log.Errorf("failed to accept connection: %v", err)
+			continue
+		}
+
+		go handleConnection(conn)
+	}
+}
+
+func handleConnection(fd int) {
+	defer unix.Close(fd)
```

Comment:
Reads and writes here are used without timeout. If the VM is stopped during this interaction, it can be a minor problem: the goroutine is blocked, and the resources are leaked. I understand that it's not critical, but we can add some handling for such situations. 

For example, we can run some heart-beating goroutine for each VM. Or we can subscribe for status from the domain manager and catch the stopping VMs, closing the corresponding connections if they are progressing at the moment...

Reply:
added timeout for the socket.